src/library/export/engineprimeexportjob.cpp

```diff
@@ -0,0 +1,560 @@
+        std::vector<djinterop::waveform_entry> externalWaveform;
+        externalWaveform.reserve(externalWaveformSize);
+        for (int64_t i = 0; i < externalWaveformSize; ++i) {
+            auto j = pWaveform->getDataSize() * i / externalWaveformSize;
+            externalWaveform.push_back({{pWaveform->getLow(j), 127},
+                    {pWaveform->getMid(j), 127},
+                    {pWaveform->getHigh(j), 127}});
+        }
+        externalTrack.set_waveform(std::move(externalWaveform));
+    } else {
+        qInfo() << "No waveform data found for track" << pTrack->getId()
+                << "(" << pTrack->getFileInfo().fileName() << ")";
+    }
+}
+
+void exportTrack(
+        const EnginePrimeExportRequest& request,
+        djinterop::database& db,
+        QHash<TrackId, int64_t>& mixxxToEnginePrimeTrackIdMap,
+        const TrackPointer pTrack,
+        std::unique_ptr<Waveform> pWaveform) {
+    // Only export supported file types.
+    if (!kSupportedFileTypes.contains(pTrack->getType())) {
+        qInfo() << "Skipping file" << pTrack->getFileInfo().fileName()
+                << "(id" << pTrack->getId() << ") as its file type"
+                << pTrack->getType() << "is not supported";
+        return;
+    }
+
+    // Copy the file, if required.
+    auto musicFileRelativePath = exportFile(request, pTrack);
+
+    // Export meta-data.
+    exportMetadata(db,
+            mixxxToEnginePrimeTrackIdMap,
+            pTrack,
+            std::move(pWaveform),
+            musicFileRelativePath);
+}
+
+void exportCrate(
+        djinterop::crate& extRootCrate,
+        QHash<TrackId, int64_t>& mixxxToEnginePrimeTrackIdMap,
+        const Crate& crate,
+        const QList<TrackId>& trackIds) {
+    // Create a new crate as a sub-crate of the top-level Mixxx crate.
+    auto extCrate = extRootCrate.create_sub_crate(crate.getName().toStdString());
+
+    // Loop through all track ids in this crate and add.
+    for (auto iter = trackIds.cbegin(); iter != trackIds.cend(); ++iter) {
+        auto trackId = *iter;
+        auto extTrackId = mixxxToEnginePrimeTrackIdMap[trackId];
+        extCrate.add_track(extTrackId);
+    }
+}
+
+} // namespace
+
+EnginePrimeExportJob::EnginePrimeExportJob(
+        QObject* parent,
+        TrackCollectionManager* pTrackCollectionManager,
+        EnginePrimeExportRequest request)
+        : QThread(parent),
+          m_pTrackCollectionManager(pTrackCollectionManager),
+          m_request{std::move(request)} {
+    // Must be collocated with the TrackCollectionManager.
+    DEBUG_ASSERT(m_pTrackCollectionManager != nullptr);
+    moveToThread(m_pTrackCollectionManager->thread());
+}
+
+void EnginePrimeExportJob::loadIds(QSet<CrateId> crateIds) {
+    // Note: this slot exists to ensure the track collection is never accessed
+    // from outside its own thread.
+    DEBUG_ASSERT(thread() == m_pTrackCollectionManager->thread());
+    QMutexLocker lock{&m_mainThreadLoadMutex};
```

Comment:
This mutex could block the UI thread for an unknown time period. Mixxx already uses too many mutexes and in the wrong way.